From 5bf2b951a0e5d8208cc992903fdd44444f80ce47 Mon Sep 17 00:00:00 2001
From: ziqic <quic_ziqic_01@quicinc.com>
Date: Thu, 20 Jun 2024 16:58:24 +0800
Subject: [PATCH] modules-load: implement parallel module loading

These changes ensure efficient module loading by parallel processing,
reducing overall load times.

Signed-off-by: Eric Curtin <ecurtin@redhat.com>
Git-commit: bdcce1b4b3003550836c9ffbb3239adae1fba75e
            a2b9ca336b075e96debf372fe93a43be3afac717
            41437d6a60281753f0a4d12249990dfcdfa7bf82
Git-repo: https://github.com/systemd/systemd/pull/33082/commits

Implement parallel module loading to optimize bootkpi.
Solve some compile errors caused by version differences.
Send some empty data to stop while loop and close threads.

Signed-off-by: ziqic <quic_ziqic_01@quicinc.com>
---
 src/modules-load/modules-load.c | 84 +++++++++++++++++++++++----------
 1 file changed, 60 insertions(+), 24 deletions(-)

diff --git a/src/modules-load/modules-load.c b/src/modules-load/modules-load.c
index eda847cb4a..7d687a29eb 100644
--- a/src/modules-load/modules-load.c
+++ b/src/modules-load/modules-load.c
@@ -3,6 +3,7 @@
 #include <errno.h>
 #include <getopt.h>
 #include <limits.h>
+#include <pthread.h>
 #include <sys/stat.h>
 
 #include "conf-files.h"
@@ -18,6 +19,8 @@
 #include "strv.h"
 #include "util.h"
 
+#define EBADF_PAIR { -EBADF, -EBADF }
+
 static char **arg_proc_cmdline_modules = NULL;
 static const char conf_file_dirs[] = CONF_PATHS_NULSTR("modules-load.d");
 
@@ -60,12 +63,35 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
         return 0;
 }
 
-static int apply_file(struct kmod_ctx *ctx, const char *path, bool ignore_enoent) {
+static void *worker_thread_module(void *arg) {
+        int sock_fd = PTR_TO_FD(arg);
+        char buffer[NAME_MAX + 1];
+        ssize_t bytes_received;
+
+        _cleanup_(kmod_unrefp) struct kmod_ctx *ctx = NULL;
+        ctx = kmod_new(NULL, NULL);
+        if (!ctx) {
+                log_error("Failed to allocate memory for kmod.");
+        }
+
+        kmod_load_resources(ctx);
+        kmod_set_log_fn(ctx, systemd_kmod_log, NULL);
+
+        int r;
+
+        while ((bytes_received = recv(sock_fd, buffer, sizeof(buffer), 0)) > 0) {
+                buffer[bytes_received] = '\0';
+                r = module_load_and_warn(ctx, buffer, true);
+        }
+
+        return NULL;
+}
+
+static int apply_file(const int *sockets, const char *path, bool ignore_enoent) {
         _cleanup_fclose_ FILE *f = NULL;
         _cleanup_free_ char *pp = NULL;
         int r;
 
-        assert(ctx);
         assert(path);
 
         r = search_and_fopen_nulstr(path, "re", NULL, conf_file_dirs, &f, &pp);
@@ -85,20 +111,20 @@ static int apply_file(struct kmod_ctx *ctx, const char *path, bool ignore_enoent
                 k = read_line(f, LONG_LINE_MAX, &line);
                 if (k < 0)
                         return log_error_errno(k, "Failed to read file '%s': %m", pp);
+
                 if (k == 0)
                         break;
 
                 l = strstrip(line);
+
                 if (isempty(l))
                         continue;
                 if (strchr(COMMENTS, *l))
                         continue;
 
-                k = module_load_and_warn(ctx, l, true);
-                if (k == -ENOENT)
-                        continue;
-                if (k < 0 && r >= 0)
-                        r = k;
+                if (send(sockets[0], l, strlen(l), 0) < 0) {
+                        log_error_errno(errno, "send");
+                }
         }
 
         return r;
@@ -159,7 +185,6 @@ static int parse_argv(int argc, char *argv[]) {
 }
 
 static int run(int argc, char *argv[]) {
-        _cleanup_(kmod_unrefp) struct kmod_ctx *ctx = NULL;
         int r, k;
 
         r = parse_argv(argc, argv);
@@ -174,20 +199,26 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 log_warning_errno(r, "Failed to parse kernel command line, ignoring: %m");
 
-        ctx = kmod_new(NULL, NULL);
-        if (!ctx) {
-                log_error("Failed to allocate memory for kmod.");
-                return -ENOMEM;
-        }
+        _cleanup_close_pair_ int seq[2] = EBADF_PAIR;
+        long ncpus = sysconf(_SC_NPROCESSORS_ONLN);
+        if (ncpus < 0)
+                log_warning_errno(r, "Failed to get number of cpus");
 
-        kmod_load_resources(ctx);
-        kmod_set_log_fn(ctx, systemd_kmod_log, NULL);
+        long max_tasks = MAX(MIN(ncpus / 2, 10), 2);
+        _cleanup_free_ pthread_t *threads = malloc(sizeof(pthread_t) * max_tasks);
+        if (socketpair(AF_UNIX, SOCK_SEQPACKET, 0, seq) == -1)
+                log_error_errno(errno, "socketpair");
+
+        // Create worker threads
+        for (int i = 0; i < max_tasks; ++i)
+                if (pthread_create(&threads[i], NULL, worker_thread_module, FD_TO_PTR(seq[1])) != 0)
+                        log_error_errno(errno, "pthread_create");
 
         r = 0;
 
         if (argc > optind) {
                 for (int i = optind; i < argc; i++) {
-                        k = apply_file(ctx, argv[i], false);
+                        k = apply_file(seq, argv[i], false);
                         if (k < 0 && r == 0)
                                 r = k;
                 }
@@ -196,13 +227,9 @@ static int run(int argc, char *argv[]) {
                 _cleanup_strv_free_ char **files = NULL;
                 char **fn, **i;
 
-                STRV_FOREACH(i, arg_proc_cmdline_modules) {
-                        k = module_load_and_warn(ctx, *i, true);
-                        if (k == -ENOENT)
-                                continue;
-                        if (k < 0 && r == 0)
-                                r = k;
-                }
+                STRV_FOREACH(i, arg_proc_cmdline_modules)
+                        if (send(seq[0], *i, strlen(*i), 0) < 0)
+                                log_error_errno(errno, "send");
 
                 k = conf_files_list_nulstr(&files, ".conf", NULL, 0, conf_file_dirs);
                 if (k < 0) {
@@ -213,12 +240,21 @@ static int run(int argc, char *argv[]) {
                 }
 
                 STRV_FOREACH(fn, files) {
-                        k = apply_file(ctx, *fn, true);
+                        k = apply_file(seq, *fn, true);
                         if (k < 0 && r == 0)
                                 r = k;
                 }
+                log_error("modules-load apply_file finish");
+        }
+
+        for (int i = 0; i < max_tasks; ++i) {
+                send(seq[0], NULL, 0, 0);
         }
 
+        // Wait for all threads to finish
+        for (int i = 0; i < max_tasks; ++i)
+                pthread_join(threads[i], NULL);
+
         return r;
 }
 
-- 
2.25.1

